src/treelearner/serial_tree_learner.h

```diff
@@ -215,6 +286,17 @@ class SerialTreeLearner: public TreeLearner {
   const Json* forced_split_json_;
   std::unique_ptr<TrainingShareStates> share_state_;
   std::unique_ptr<CostEfficientGradientBoosting> cegb_;
+  /*! \brief whether numerical features contain any nan values, used for linear model */
+  std::vector<int8_t> contains_nan_;
+  /*! whether any numerical feature contains a nan value, used for linear model */
+  bool any_nan_;
+  /*! \brief map dataset to leaves, used for linear model */
+  std::vector<int> leaf_map_;
+  /*! \brief temporary storage for calculating linear model */
+  std::vector<std::vector<float>> XTHX_;
+  std::vector<std::vector<float>> XTg_;
+  std::vector<std::vector<std::vector<float>>> XTHX_by_thread_;
+  std::vector<std::vector<std::vector<float>>> XTg_by_thread_;
```

Comment:
above temporary storage could be mutable, so that we can make tree_learner const.

Reply:
Done.